parlai/agents/transformer/polyencoder.py

```diff
@@ -593,17 +496,180 @@ def forward(
         :param cand_rep:
             encoded representation of the candidates
         """
-        if ctxt_tokens is not None or ctxt_image is not None or cand_tokens is not None:
-            return self.encode(
-                ctxt_tokens=ctxt_tokens, ctxt_image=ctxt_image, cand_tokens=cand_tokens
-            )
+        if len(ctxt_inputs) > 0 or cand_tokens is not None:
+            return self.encode(cand_tokens=cand_tokens, **ctxt_inputs)
         elif (
             ctxt_rep is not None and ctxt_rep_mask is not None and cand_rep is not None
         ):
             return self.score(ctxt_rep, ctxt_rep_mask, cand_rep)
         raise Exception('Unsupported operation')
 
 
+class ImagePolyencoderAgent(PolyencoderAgent):
+    """
+    Poly-encoder Agent that ingests image features.
+
+    Agent that allows encoding image features and adding or concatenating them to the
+    context encoding.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add command-line arguments specifically for this agent.
+        """
+        super(ImagePolyencoderAgent, cls).add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('Image Encoder Args')
+        agent.add_argument(
+            '--polyencoder-image-encoder-num-layers',
+            type=int,
+            default=1,
+            help='Number of linear layers to encode image features with in the context',
+        )
+        agent.add_argument(
+            '--polyencoder-image-features-dim',
+            type=int,
+            default=DEFAULT_IMAGE_FEATURES_DIM,
+            help='For passing in image features of the given dim in the context',
+        )
+        agent.add_argument(
+            '--polyencoder-image-combination-mode',
+            type=str,
+            default='prepend',
+            choices=['add', 'append', 'prepend'],
+            help='How to combine image embedding (if used) with context embedding',
+        )
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.image_features_dim = opt.get(
+            'polyencoder_image_features_dim', DEFAULT_IMAGE_FEATURES_DIM
+        )
+
+    def build_model(self, states=None):
+        """
+        Return built model.
+        """
+        return ImagePolyencoderModule(self.opt, self.dict, self.NULL_IDX)
+
+    def batchify(self, obs_batch: List[Message], sort: bool = False) -> Batch:
+        """
+        Override to handle image features.
+        """
+        batch = super().batchify(obs_batch, sort)
+
+        def _process_features(features: torch.Tensor) -> torch.Tensor:
```

Comment:
perhaps `_process_image_features` to make it more clear?

Reply:
Yes - changed